scripts/cnode-helper-scripts/cntools.library

```diff
@@ -61,22 +61,48 @@ logln() {
 #                             INFO   : normal output printed to tty and logged
 #                             ACTION : e.g cardano-cli executions etc, logged but not printed to tty
 #                             ERROR  : stderr output and error messages
+#             : newline    >  Add a newline at the end for tty output (default true) 
 #             : message    >  The message to print/log
 println() {  
   sleep 0.05 # hack, sleep 50ms before each print to preserve order of execution
   local log_level=$1
   shift
+  local newline="\n"
+  if [[ $1 =~ true|false ]]; then
+    [[ $1 = false ]] && newline=""
+    shift
+  fi
   case $log_level in
-    OFF) printf '%b\n' "$@" >&6 ;;
+    OFF) printf "%b${newline}" "$@" >&6 ;;
     LOG) logln "DEBUG" "$@" ;;
-    DEBUG) printf '%b\n' "$@" >&6; logln "DEBUG" "$@" ;;
-    INFO) printf '%b\n' "$@" ;;
+    DEBUG) printf "%b${newline}" "$@" >&6; logln "DEBUG" "$@" ;;
+    INFO) printf "%b${newline}" "$@" ;;
     ACTION) logln "ACTION" "$@" ;;
-    ERROR) printf '%b\n' "$@" >&2 ;;
-    *) printf '%b\n' "${log_level}"; [[ $# -ge 1 ]] && printf '%s\n' "$@" ;;
+    ERROR) printf "%b${newline}" "$@" >&2 ;;
+    *) printf "%b${newline}" "${log_level}"; [[ $# -ge 1 ]] && printf "%b${newline}" "$@" ;;
   esac
 }
 
+# Command     : ask [variable name] [log] [question]
+# Description : read input from stdin and save response into provided variable name while also logging response
+# Parameters  : variable name  >  the name of the variable to save users response into
+#             : log            >  [true|false] log question (default: true)
+#             : question       >  what to ask user to input
+ask() {
```

Comment:
Should we move this to env?

Reply:
println() and ask() are both coded a bit special due to CNTools file descriptor redirect change.
Functions print output to FD 6 (normal stdout in CNTools instead of 1) to **not** automatically trigger logging. Then if logging should be done(variable to function), this is triggered as a separate action.

Other scripts don't utilize these special FD redirects.

Reply:
Actually, I think we can move core of function to env, at least for ask.